lib/dry/core/memoizable.rb

```diff
@@ -44,13 +44,13 @@ def define_memoizable_names!
             meth = klass.instance_method(name)
 
             if meth.parameters.size > 0
-              define_method(name) do |*args|
-                name_with_args = :"#{name}_#{args.hash}"
+              define_method(name) do |*args, &block|
```

Comment:
Hmm. How about we actually look if a method expects a block? Otherwise we're redefining every method and adding support for blocks even when the original method doesn't really accept blocks. I mean, I know this can be handled implicitly in Ruby but maybe we should diverge from this and just establish (and document) a convention that if you want a memoizable method with a block, then its signature should include a block argument.

I also wonder if this could be rewritten in such a way that we maintain arguments exactly as in the original method via class_eval and going through parameters 🤔

Reply:
@solnic What's your objective? Minimize confusion, performance? Just so I know what we're optimizing towards (⌒‿⌒)

Reply:
Ah, I understand what you mean now. You want the memorized method to not only behave in the same way as the original one but also to return the same stats as the original, i.e trough method(:mymethod).parameters. 

Makes sense. I’ll look in to it.

Reply:
We could use, *cough* *cough* `module_eval`: (tested in Ruby 2.5.0)

``` ruby
require "rspec"

class Memoizer < Module
  def initialize(klass, names)
    names.each do |name|
      set(klass, klass.instance_method(name))
    end
  end

  private

  def to_declaration(params)
    params.map do |type, name|
      case type
      when :req
        name.to_s
      when :opt
        if (name = params.to_h[:rest])
          "*#{name}"
        else
          "*args"
        end
      when :rest
        "*#{name}"
      when :keyreq
        "#{name}:"
      when :key
        if (name = params.to_h[:keyrest])
          "**#{name}"
        else
          "**kwargs"
        end
      when :keyrest
        "**#{name}"
      when :block
        "&#{name}"
      end
    end.uniq.compact
  end

  def to_super(params)
    params.map do |param|
      param[-1] == ":" ? "#{param} #{param[0...-1]}" : param
    end
  end

  def set(klass, method)
    params = method.parameters
    declaration_params = to_declaration(params)
    super_params = to_super(declaration_params)

    module_eval <<~RUBY, __FILE__, __LINE__ + 1
      def #{method.name}(#{declaration_params.join(", ")})
        super(#{super_params.join(", ")})
      end
    RUBY
  end
end

module Memo
  module ClassInterface
    def memoize(*methods)
      prepend(Memoizer.new(self, methods))
    end
  end

  def self.included(klass)
    super
    klass.extend(ClassInterface)
  end
end

class Test
  include Memo

  def test1(arg1, *args, kwarg1:, kwarg2: "default3", **kwargs, &block)
    # NOP
  end

  def test2(arg1, *args, &block)
    # NOP
  end

  def test3(&block)
    # NOP
  end

  def test4
    # NOP
  end

  memoize :test1, :test2, :test3, :test4
end

RSpec.shared_examples 'a memoized method' do
  let(:new_meth) { described_class }
  let(:old_meth) { new_meth.super_method }

  describe "new != old" do
    subject { new_meth }
    it { is_expected.not_to eq(old_meth) }
  end

  describe "new.arity == old.arity" do
    subject { new_meth.arity }
    it { is_expected.to eq(old_meth.arity) }
  end

  describe "new.name == old.name" do
    subject { new_meth.name }
    it { is_expected.to eq(old_meth.name) }
  end
end

RSpec.describe Test.new do
  let(:block) { -> { } }

  describe described_class.method(:test1) do
    it_behaves_like 'a memoized method'

    it "does not raise an error" do
      described_class.call("a", kwarg1: "world", other: "test", &block)
    end
  end

  describe described_class.method(:test2) do
    it_behaves_like 'a memoized method'

    it "does not raise an error" do
      described_class.call("a", &block)
    end
  end

  describe described_class.method(:test3) do
    it_behaves_like 'a memoized method'

    it "does not raise an error" do
      described_class.call(&block)
    end
  end

  describe described_class.method(:test4) do
    it_behaves_like 'a memoized method'

    it "does not raise an error" do
      described_class.call
    end
  end
end
```

Reply:
@oleander this is *really nice*! Please give this a shot 🙂